stl/inc/algorithm

```diff
@@ -6080,6 +6096,92 @@ namespace ranges {
 #endif // _HAS_CXX20
 #endif // _HAS_CXX17
 
+// Batched random integer generation for shuffle optimization.
+// From Nevin Brackett-Rozinsky and Daniel Lemire, "Batched Ranged Random Integer Generation",
+// Software: Practice and Experience 55(1), 2025.
+//
+// This algorithm extracts multiple bounded random integers from a single 64-bit random word,
+// using only multiplication (no division) in the common case.
+
+template <class _Diff, class _Urng>
+struct _Batched_rng_from_urng {
+    // Batched random generation is only beneficial for 64-bit RNGs with full range.
+    // It requires the RNG to produce values in [0, 2^64 - 1].
+    using _Urng_result = _Invoke_result_t<_Urng&>;
+
+    static constexpr bool _Has_full_64bit_range = sizeof(_Urng_result) >= sizeof(uint64_t)
+                                               && is_unsigned_v<_Urng_result> && (_Urng::min) () == 0
+                                               && (_Urng::max) () == (numeric_limits<uint64_t>::max) ();

```

Comment:
Note that `is_unsigned_v<T>` can only be `true` for integer types. I think on MSVC ABI only `unsigned long long` (and its cv-qualified versions) satisfies such conditions.

Reply:
Done.